Fix categorical random shape

canyon289 opened the pull request:
Hello,
This pull request is for https://github.com/pymc-devs/pymc3/issues/3035 and is very much in WIP.

To summarize I tried to hit two objectives per @junpenglao request, one was removing duplication of random draws from categorical, the other was address the shape.

For removing duplication
I created a function in distribution called *random_choice* and pointed the previous methods to it. All tests are passing in this structure.

Questions I have
Is this name good or should it be called categorical_random_choice?
Is distributions.py a good place for it?


For the shape comment
I believe there is an expectation that if a (10,5) size is requested, the method returns a (10,5) array. 
These tests specifically seem to test for that behavior.
`pymc3.tests.test_distributions_random.BaseTestCases.BaseTestCase#test_scalar_parameter_shape`
`pymc3.tests.test_distributions_random.BaseTestCases.BaseTestCase#test_scalar_shape`
Given that I don't understand how this test reconciles with the original issue, but this is not to suggest the request was wrong. I'm very much at the edge of my understanding and feel like I'm missing something.
How should I proceed?

I've attached some supporting screenshots that may help. Thanks for the chance to participate and leanr

![image](https://user-images.githubusercontent.com/7213793/41921156-178f89dc-7917-11e8-84c2-3f858f756488.png)
![image](https://user-images.githubusercontent.com/7213793/41921376-96d0f712-7917-11e8-8c8a-ee70a6f64ec1.png)

junpenglao commented:
> Is this name good or should it be called categorical_random_choice?

I am fine with it

> Is distributions.py a good place for it?

distribution.dist_math might be better

junpenglao commented:
Also rebase needed.

canyon289 commented:
@junpenglao Currently running linting and coverage overnight but do let me know if anything is missing.

I still do have a question about the comment on the shapes. Given the test_scalar_parameter_shape test does it still make sense to have random_choice return an (x,1) shape array?

Thanks for the quick comment

canyon289 commented:
Some tests are failing. I'll check them out

canyon289 commented:
One last thing, I still need to run the examples in docs/source/notebooks, but wanted to make sure my implementation at the code level was ok before proceeding.

junpenglao commented:
Cool. 
Need a test to make sure the case in https://github.com/pymc-devs/pymc3/issues/3035 is now return array with the correct shape.

twiecki commented on pymc3/distributions/dist_math.py:
```diff
@@ -261,3 +261,29 @@ def impl(self, x):
+        samples = np.row_stack([np.random.choice(k, p=p_) for p_ in p])
+    else:
+        samples = np.random.choice(k, p=p, size=size)
+    return samples
```
missing new-line

canyon289 commented:
@junpenglao and @twiecki thanks for the feedback

In regards to the shape, I think an input size of (10,5) matrix with an output size of (10,) would break the below but I could be mistaken. Is my understanding correct? Below is a screenshot as well showing how the test expects random_choice to output the same size array as the input.

Lastly I'll be away for the next week but can finish this when I get back!

https://github.com/pymc-devs/pymc3/blob/master/pymc3/tests/test_distributions_random.py#L148-L168

![image](https://user-images.githubusercontent.com/7213793/42012993-ef162954-7a4f-11e8-9917-3e7a861ed520.png)

junpenglao commented:
Yes you should change that test for Categorical - the shape of the output from Categorical random is not the same as parameters.

canyon289 commented:
Hello, I'm back and will be working on this again this weekend! Just wanted to give a heads up

canyon289 commented on pymc3/distributions/discrete.py:
```diff
@@ -710,19 +710,9 @@ def __init__(self, p, *args, **kwargs):
-            return np.asarray(np.random.choice(k, p=p, size=size))
-
     def random(self, point=None, size=None):
         p, k = draw_values([self.p, self.k], point=point, size=size)
```
@junpenglao @twiecki  Just for my own understanding why is draw_values called here? Why can't we just use `generate_samples` directly?

canyon289 replied:
I realize this may be a really basic question, but I'm biasing towards asking rather than assuming as I learn more about pymc3 :)

ColCarroll replied:
sorry, i think this was a question for me: `draw_values` is generally to get the parameters of a distribution to the right shape (`size=size`), with the right conditioning (`point=point`).  The `generate_samples` function actually draws new samples for an RV. 

If you trace the code, `draw_values` will call `.random` on each of the variables, which will (usually) end up at `generate_samples`.  This is how we traverse back up the computation graph.

canyon289 committed:
Modify random_choice to return correct shape

canyon289 commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -402,6 +402,17 @@ class TestCategorical(BaseTestCases.BaseTestCase):
+    def test_scalar_parameter_shape(self):
+        rv = self.get_random_variable(None)
+        # for size, expected in ((None, (1,)), (5, (5, 1)), ((4, 5), (4, 1))):
+        for size, expected in (((4, 5), (4, 1)),):
```
So while I fixed the original inciting case in the issue, I don't know enough about all the possible ways that categorical.random could be called to know how to modify these tests correctly.

The original example passes `p` as a parameter to `dist.random` and that case is handled
This test passes `size` to `dist.random` and I don't understand how that should be handled by `Categorical.random` or if this ever happens during nominal use of this library.

Any guidance would be appreciated and thanks for bearing with me while I learn the ropes.
![image](https://user-images.githubusercontent.com/7213793/42424627-1c1855b8-82c4-11e8-9bf2-aac5b59077c2.png)

ColCarroll replied:
Just to be clear, in this example, you expected `v.shape` to be `(10, 5)`, right?  I tried to get `draw_values` and `generate_samples` to return something shaped roughly the following way:

1. broadcast the parameters together using `np.broadcast` to get their shape. Call it `d1`.
2. look at the `shape` argument passed to the random variable, and call it `d2`.  Try to combine `d1` and `d2` in a sensible way, and throw an exception instead of guessing _too_ much.  This is the bulk of the `draw_values` code.  We'll say `d3` is the combination of `d1` and `d2`
3. return something of shape `size x d3`

Explicitly, consider `x = pm.Normal.dist(mu=np.random.rand(5), sd=1, shape=(2, 5)).random(size=10)

1. The size of `mu` is `(5,)`, and `sd` is `()`, and broadcasting them gives `(5,)`
2. The shape argument (`(2, 5)`) fits nicely with the broadcast shape, so `d3` above is `(2, 5)`.  Note that `shape=(2, 4)` throws an exception.
3. We return an array of shape `(10, 2, 5)`, since the size argument was 10.

I hope this helps!

canyon289 replied:
Hey @ColCarroll 
I believe v.shape should be (10,1) per the issue @junpenglao filed? https://github.com/pymc-devs/pymc3/issues/3035

I've read your comments and want to read them a couple more times before asking any other follow up questions. This is very helpful!

canyon289 replied:
@ColCarroll and @junpenglao 

Would either of you mind if I put the deduplication of `random_choice` on another pull request so it can be merged independently of the shape issue? All tests were passing for that portion of the code.

I ask because I think it'll streamline the code changes and reduce the risk of conflicts later.

junpenglao replied:
Sure thing. Which PR is that?

canyon289 replied:
Just created it here. Pending CI tests at time of writing
https://github.com/pymc-devs/pymc3/pull/3084

canyon289 commented:
@junpenglao and @ColCarroll Starting a new thread since I jumbled the last one. I also updated the commit tree with code that only relevant to the discussion around shape.

To summarize
I fixed the original issue that @junpenglao referenced in this issue, where a 2 dimensional array is passed to `p`, pymc3 should return a (n,1) column vector for Categorical distributions
https://github.com/pymc-devs/pymc3/issues/3035

To directly address @ColCarroll's statement. I do *not* expect `v.shape` to be (10,5) due to @junpenglao's original ticket, I expect it to return a (10,1) column vector.

However I have no idea if this change breaks people's code. I certainly broke a number of tests. The tests all ensure that the correct `size` is returned when a`size` argument is passed. 

I tried reading a bunch of closed issues but I ended up a little more confused than I originally started.
https://github.com/pymc-devs/pymc3/issues/2614
https://github.com/pymc-devs/pymc3/pull/2623
https://github.com/pymc-devs/pymc3/pull/2876
https://github.com/pymc-devs/pymc3/pull/2979

So starting off with my first question
@ColCarroll When you say parameter in Step 1 above, in this case the parameter would be `p` and the first step is for np.broadcast to get `p`'s shape?

Lastly, if it becomes too much to answer my questions please let me know. My intention is to help contribute to pymc3 but if it takes 5 hours of questions from me to get 1 hour of work done then it's may have ended up detracting from you guys :( I truly do appreciate the help thus far.

Thank you

junpenglao commented:
Looking at the failed test, I think you are getting really close :-)

Currently, most of the test fail is matching (n, ) with (n, 1). I dont have a strong feeling of this, I guess if you modify it to return a (n, ) is fine as well. Not sure which one makes more sense tho... I asked twitter https://twitter.com/junpenglao/status/1018242975682453504

the test mismatch of (4, 5) and (4, 1) you should just go ahead and modify the test.

the test mismatch of (5, 10, 5) and (50, 5) would need some investigation.

canyon289 commented:
@ColCarroll @junpenglao 
I modified Categorical.random back to (n,) output and all tests pass. In that situation all tests pass and I could use a review

junpenglao commented:
Did you implement a new test for this?

canyon289 commented:
I did not but should have, I'll change this back to WIP and do that sometime before Sunday, My apologies

canyon289 committed:
Add test for categorical distribution shape

canyon289 committed:
Remove unused arrays in test

canyon289 commented:
@junpenglao I think I'm ready for what I hope is the final code review! Hopefully this is the one that makes it

junpenglao commented:
Awesome! One less shape issue ;-)
